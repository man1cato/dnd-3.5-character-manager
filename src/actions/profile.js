import database from '../firebase/firebase'


//SET PROFILE IN STORE
export const setProfile = (id, profile) => ({
    type: 'SET_PROFILE',
    id,
    profile
});

//READ PROFILE FROM FIREBASE
export const startSetProfile = (id) => {
    return (dispatch, getState) => {
        const uid = getState().auth.uid;
        return database.ref(`users/${uid}/profiles/${id}`).once('value').then((snapshot) => {
            const profile = snapshot.val();
            dispatch(setProfile(id, profile));
        });
    };
};

//CREATE PROFILE IN FIREBASE
export const startCreateProfile = (profile) => {    
    return (dispatch, getState) => {
        const uid = getState().auth.uid       
        return database.ref(`users/${uid}/profiles`).push(profile).then((ref) => {
            dispatch(setProfile(ref.key, profile))
        })
    }
}


//UPDATE PROFILE IN STORE
export const editProfile = (updates) => ({
    type: 'EDIT_PROFILE',
    updates
});

//UPDATE PROFILE IN FIREBASE
export const startEditProfile = (id, updates) => {
    return (dispatch, getState) => {
        const uid = getState().auth.uid;
        return database.ref(`users/${uid}/profiles/${id}`).update(updates).then(() => {
            dispatch(editProfile(updates));
        });
    };
};

//DELETE PROFILE FROM STORE
export const deleteProfile = () => ({
    type: 'DELETE_PROFILE'
})

//LEVEL UP CHARACTER IN STORE
// export const levelUp = (character) => ({
//     type: 'LEVEL_UP',
//     character
// });
