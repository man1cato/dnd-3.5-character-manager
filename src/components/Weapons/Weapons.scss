@import '../../styles/base/settings';


.weapon {
   display: flex;
   flex-direction: column;
   text-align: center;
   margin-bottom: $m-size;
}

.weapon__features {
   font-size: $font-size-s;  
   padding-bottom: $xs-size;
}

.weapon__stats {
   display: flex;
   align-items: center;
   justify-content: space-evenly;
   > * {
      flex: 1 1 25%;
   }
   > :not(:last-child) {
      border-right: solid 1px $dark-grey;
   }
}
