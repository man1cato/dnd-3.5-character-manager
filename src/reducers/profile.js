export default (state = {}, action) => {
    switch (action.type) {
        case 'SET_PROFILE':
            return {
                id: action.id,
                ...action.fields
            };
        case 'EDIT_PROFILE':
            return {
                ...state,
                ...action.updates
            };
        default:
            return state;
    }
};
