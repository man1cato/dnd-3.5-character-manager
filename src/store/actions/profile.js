import database from '../../firebase/firebase'
import { startGetProfiles } from './profiles'

//SET SELECTED PROFILE IN STORE
export const setProfile = (id, profile) => ({
    type: 'SET_PROFILE',
    id,
    profile
})

//READ PROFILE FROM FIREBASE
export const startSetProfile = (uid, id) => {
    return (dispatch) => {
        return database.ref(`users/${uid}/profiles/${id}`).once('value').then((snapshot) => {
            const profile = snapshot.val()
            dispatch(setProfile(id, profile))
        })
    }
}

//CREATE PROFILE IN FIREBASE
export const startCreateProfile = (profile) => {    
    return (dispatch, getState) => {
        const uid = getState().auth.uid       
        return database.ref(`users/${uid}/profiles`).push(profile).then((ref) => {
            dispatch(setProfile(ref.key, profile))
            dispatch(startGetProfiles(uid))
            localStorage.setItem('selectedCharacterId', ref.key)
        })
    }
}


//UPDATE PROFILE IN STORE
export const editProfile = (updates) => ({
    type: 'EDIT_PROFILE',
    updates
})

//UPDATE PROFILE IN FIREBASE
export const startEditProfile = (id, updates) => {
    return (dispatch, getState) => {
        const uid = getState().auth.uid
        return database.ref(`users/${uid}/profiles/${id}`).update(updates).then(() => {
            dispatch(editProfile(updates))
        })
    }
}

//REMOVE PROFILE FROM STORE
export const removeProfile = () => ({
    type: 'REMOVE_PROFILE'
})

//REMOVE PROFILE FROM FIREBASE
// export const startRemoveProfile = () => {
//     return (dispatch) => {
//         return database.ref(`users/${uid}/profiles/${id}`).remove().then(() => {
//             dispatch(removeProfile())
//         })
//     }
// }

//LEVEL UP CHARACTER IN STORE
// export const levelUp = (character) => ({
//     type: 'LEVEL_UP',
//     character
// });
