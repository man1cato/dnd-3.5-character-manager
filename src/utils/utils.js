import _ from 'lodash'


export const convertTextToArray = (string) => _.compact(string.split(/\n/))

export const findItemById = (storeItemArray, itemId) => storeItemArray.find((item) => item.id === itemId)

export const apiObjectToArray = (obj) => _.orderBy(Object.keys(obj).map((id) => ({
	id,
	...obj[id]
})), ['name'], ['asc'])