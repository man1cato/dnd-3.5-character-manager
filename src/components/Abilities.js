import React, { Fragment } from 'react'
import _ from 'lodash'
import { calcAbilityMod } from '../utils/utils'


const Abilities = ({ abilities, handleChange }) => (
	<div className="grid--abilities">
		<h5 className="grid__col1">Ability</h5>
		<h5 className="grid__col2">Score</h5>
		<h5 className="grid__col3">Temp Score</h5>
		<h5 className="grid__col4"> Mod</h5>

		{_.map(_.keys(abilities), (ability, i) => (
			<Fragment key={i}>
				<div className="grid__col1">{ability.toUpperCase()}</div>
				<div className="grid__col2">{abilities[ability].score}</div>
				<input
					className="grid__col3"
					id={ability}
					name="abilities"
					type="number"
					value={abilities[ability].tempScore}
					onChange={handleChange}
				/>
				<div className="grid__col4">
					{!!abilities[ability].tempScore ? calcAbilityMod(abilities[ability].tempScore) : calcAbilityMod(abilities[ability].score)}
				</div>
			</Fragment>
		))}
	</div>
)


export default Abilities