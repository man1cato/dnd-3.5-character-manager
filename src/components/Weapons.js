import React from 'react'
import { connect } from 'react-redux'
import { apiObjectToArray } from '../utils/utils'


export const Weapons = (props) => (
	<>
		{props.weaponSet.map((weaponId) => {
			const weapon = props.weapons.find((weapon) => weapon.id === weaponId)

			return (
				<div className="grid--weapons" key={weaponId}>
					<h4 className=" grid--weapons__name">{weapon.name}</h4>
					<div className="grid--weapons__features">{weapon.weaponType}, {weapon.encumbrance}, {weapon.damageType}</div>

					<h5 className="grid--weapons__track">Rng</h5>
					{/* <h5 className="grid--weapons__track">Atk</h5> */}
					<h5 className="grid--weapons__track">Dmg</h5>
					<h5>Crit</h5>

					<div className="grid--weapons__track">{weapon.range}</div>
					{/* <div className="grid--weapons__track">+{
						(weapon.encumbrance === "Ranged" && props.rangedBonus)
						|| (weapon.encumbrance === "Unarmed" && props.grappleBonus)
						|| props.meleeBonus
					}</div> */}
					<div className="grid--weapons__track">{weapon.damageM}</div>
					<div>{weapon.critical}</div>
				</div>
			)
		})}
	</>
)

const mapStateToProps = (state) => ({
   weapons: apiObjectToArray(state.items).filter((item) => item.category === "Weapon")
})

export default connect(mapStateToProps)(Weapons)