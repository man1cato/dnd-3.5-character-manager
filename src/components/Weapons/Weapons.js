import React from 'react'
import { connect } from 'react-redux'

import { apiObjectToArray } from '../../utils/utils'
import './Weapons.scss'


export const Weapons = ({ weapons, equippedWeapons, attackBonuses }) => (
	<>
		{equippedWeapons.map(weaponId => {
			const weapon = weapons.find((weapon) => weapon.id === weaponId)
			return (
				<div className="weapon" key={weaponId}>
					<h4>{weapon.name}</h4>
					<div className="weapon__features">{weapon.weaponType}, {weapon.encumbrance}, {weapon.damageType}</div>

					<div className="weapon__stats">
						<div>
							<h5>Rng</h5>
							<div>{weapon.range}</div>
						</div>
						{/* <div>
							<h5>Atk</h5>
							<div>+{
								(weapon.encumbrance === "Ranged" && attackBonuses.rangedBonus)
								|| (weapon.encumbrance === "Unarmed" && attackBonuses.grappleBonus)
								|| attackBonuses.meleeBonus
							}</div>
						</div> */}
						<div>
							<h5>Dmg</h5>
							<div>{weapon.damageM}</div>
						</div>
						<div>
							<h5>Crit</h5>
							<div>{weapon.critical}</div>
						</div>
					</div>
				</div>
			)
		})}
	</>
)


const mapStateToProps = state => ({
   weapons: apiObjectToArray(state.items).filter((item) => item.category === "Weapon")
})

export default connect(mapStateToProps)(Weapons)