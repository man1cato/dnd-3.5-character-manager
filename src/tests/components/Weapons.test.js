import React from 'react'
import { shallow } from 'enzyme'
import fs from 'fs'

import { Weapons } from '../../components/Weapons'
import profile from '../fixtures/profile'
import {apiObjectToArray} from '../../utils/utils'

let props
beforeAll(async () => {
   const api = await fs.promises.readFile('src/tests/fixtures/api.json')
   const items = JSON.parse(api).items

   props = { 
       weaponSet: profile.weaponSet,
       weapons: apiObjectToArray(items).filter((item) => item.category === "Weapon")
   }
})

test('should render weapons with profile data', async () => {
    const wrapper = shallow(<Weapons {...props} />)
    expect(wrapper).toMatchSnapshot()
})
