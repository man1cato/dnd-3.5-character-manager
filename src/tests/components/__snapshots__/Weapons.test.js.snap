// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render weapons with profile data 1`] = `
<Fragment>
  <div
    className="grid--weapons"
    key="recOvuth9QArkI54l"
  >
    <h4
      className=" grid--weapons__name"
    >
      Bastard Sword
    </h4>
    <div
      className="grid--weapons__features"
    >
      Exotic
      , 
      One-Handed
      , 
      Slashing
    </div>
    <h5
      className="grid--weapons__track"
    >
      Rng
    </h5>
    <h5
      className="grid--weapons__track"
    >
      Dmg
    </h5>
    <h5>
      Crit
    </h5>
    <div
      className="grid--weapons__track"
    >
      —
    </div>
    <div
      className="grid--weapons__track"
    >
      1d10
    </div>
    <div>
      19–20/×2
    </div>
  </div>
</Fragment>
`;
